src/coreclr/System.Private.CoreLib/src/System/Runtime/DependentHandle.cs

```diff
@@ -0,0 +1,259 @@
+        internal void UnsafeSetDependent(object? dependent)
+        {
+            nSetSecondary(_handle, dependent);
+        }
+
+        /// <summary>
+        /// Retrieves the values of both <see cref="Target"/> and <see cref="Dependent"/>, if available.
+        /// </summary>
+        /// <param name="dependent">The dependent instance, if available.</param>
+        /// <returns>The values of <see cref="Target"/> and <see cref="Dependent"/>.</returns>
+        /// <remarks>
+        /// This method mirrors <see cref="GetTargetAndDependent"/>, but without the allocation check.
+        /// The signature is also kept the same as the one for the internal call, to improve the codegen.
+        /// </remarks>
+        internal object? UnsafeGetTargetAndDependent(out object? dependent)
+        {
+            return nGetPrimaryAndSecondary(_handle, out dependent);
+        }
+
+        /// <inheritdoc cref="IDisposable.Dispose"/>
+        public void Dispose()
+        {
+            // Forces dependentHandle back to non-allocated state (if not already there)
+            // and frees the handle if needed.
+
+            if (_handle != IntPtr.Zero)
+            {
+                IntPtr handle = _handle;
+                _handle = IntPtr.Zero;
+                nFree(handle);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.InternalCall)]
+        private static extern IntPtr nInitialize(object? primary, object? secondary);
+
+#if DEBUG
+        [MethodImpl(MethodImplOptions.InternalCall)]
+        private static extern object? nGetPrimary(IntPtr dependentHandle);
+#else
+        private static unsafe object? nGetPrimary(IntPtr dependentHandle)
+        {
+            // This optimization is the same that is used in GCHandle in RELEASE mode.
+            // This is not used in DEBUG builds as the runtime performs additional checks.
+            // This conversion works because on CoreCLR, DependentHandle is really just a
+            // OBJECTHANDLE value, the same that GCHandle uses. They are defined as follows:
+            //
+            // struct OBJECTHANDLE__
+            // {
+            //     void* unused;
+            // };
+            //
+            // typedef struct OBJECTHANDLE__* OBJECTHANDLE;
+            //
+            // The way this works is that the GC will allocate a special region of memory where
+            // the object reference will be stored (that is, an object* pointer), and the GC will
+            // take care of updating it whenever a compacting collection occurs and the object is
+            // moved. The code below is just doing an indirect memory access to read the pointer
+            // value stored in this memory region, and then converting that to an object reference.
+            // Note that Unsafe.As<IntPtr, object> returns a ref object and does an implicit read.
+            // The (nint) cast is needed to avoid emitting a call to IntPtr::op_Explicit(nint), which
```

Comment:
Could you please add the same tweak to the GCHandle so that the two stay in sync?

Reply:
Fixed in 0d669c7b2b8db6d442685c36a36186e41b1fee98.